lib/elixir/lib/kernel/typespec.ex

```diff
@@ -475,14 +507,94 @@ defmodule Kernel.Typespec do
   def type_to_signature({:::, _, [{name, _, args}, _]}) when is_atom(name),
     do: {name, length(args)}
 
+  defp split_spec({:when, _, [{:::, _, [spec, return]}, guard]}, _default) do
+    {spec, return, guard}
+  end
+
+  defp split_spec({:when, _, [spec, guard]}, default) do
+    {spec, default, guard}
+  end
+
+  defp split_spec({:::, _, [spec, return]}, _default) do
+    {spec, return, []}
+  end
+
+  defp split_spec(spec, default) do
+    {spec, default, []}
+  end
+
+  defp ensure_not_default({:\\, _, [_, _]}) do
+    raise ArgumentError, "default arguments \\\\ not supported in @callback/@macrocallback"
+  end
+
+  defp ensure_not_default(_), do: :ok
+
   ## Macro callbacks
 
   @doc false
+  def defspec(:callback, expr, caller) do
+    {spec, _, _} = split_spec(expr, quote(do: term))
```

Comment:
Since we don't do those checks for `@spec`, we should not do it for `@callback`. Or, alternatively, we should do it on all of them. My suggestion: keep this defspec as tiny as possible (don't split the spec nor check the arguments here). Store macrocallback still as macrocallback and process those just later on here:

https://github.com/antipax/elixir/blob/behaviour-enhancements-3465/lib/elixir/lib/kernel/typespec.ex#L662

In other words, inside that function you will check if `kind == :macrocallback` and convert the kind to callback, change the name and add the extra argument (`Macro.Env.t`) there.